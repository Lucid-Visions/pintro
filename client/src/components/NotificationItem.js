import React from "react";
import {
  View,
  StyleSheet,
  Text,
  TouchableOpacity,
  Image,
  AsyncStorage,
  TouchableHighlight,
} from "react-native";
import FollowRequestModal from "../components/followRequestModal";
import ConnectModal from "../components/ConnectModal";
import Swipeable from "react-native-swipeable-row";

import TimeAgo from "javascript-time-ago";
import en from "javascript-time-ago/locale/en";
TimeAgo.addLocale(en);
const timeAgo = new TimeAgo();

/**
 * Notification screen
 */

class NotificationItemComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isSelected: false,
      followModal: false,
      accepted: false,
    };

    this.toggleFollowModal = this.toggleFollowModal.bind(this);
    this.respond = this.respond.bind(this);
  }

  addToCircle = async (circle) => {
    const token = await AsyncStorage.getItem("token");
    const myHeaders = new Headers();
    myHeaders.append("Content-Type", "application/json");
    myHeaders.append("Authorization", token);
    const raw = JSON.stringify({
      follower: this.props.data.follower._id,
      circle,
    });

    const requestOptions = {
      method: "PATCH",
      headers: myHeaders,
      body: raw,
    };

    const response = await fetch(
      `http://${env.host}:${env.port}/api/v1/user/circle`,
      requestOptions
    ).catch((error) => console.log("error", error));

    const responseJson = await response.json();

    if (responseJson.error) {
      console.log(responseJson.error);
    }

    this.toggleFollowModal(false);
  };

  // respond to follow request
  async respond(x) {
    this.setState({
      accepted: x,
    });
    var token = await AsyncStorage.getItem("token");
    var myHeaders = new Headers();
    myHeaders.append("Content-Type", "application/json");
    myHeaders.append("Authorization", token);

    var raw = JSON.stringify({
      follower: this.props.data.follower._id,
      response: x,
    });

    var requestOptions = {
      method: "POST",
      headers: myHeaders,
      body: raw,
    };

    let response = await fetch(
      `http://${env.host}:${env.port}/api/v1/connection/respond`,
      requestOptions
    ).catch((error) => console.log("error", error));
    if (!x) this.toggleFollowModal(false);
    this.props.refresh();
  }

  getRequestButtons = () => {
    if (
      this.props.notification_type == "relationship" &&
      this.props.data.status == "PENDING"
    ) {
      return (
        <View style={{ flexDirection: "row", ...styles.requestButtons }}>
          {this.requestActionButton(
            require("../assets/blackCross.png"),
            () => {
              this.setState({ followModal: false });
              this.respond(false);
            },
            { borderWidth: 1 }
          )}
          {this.requestActionButton(
            require("../assets/WhiteTick.png"),
            () => this.respond(true),
            { backgroundColor: "black", marginLeft: 12 }
          )}
        </View>
      );
    } else {
      return null;
    }
  };

  requestActionButton(src, onPress, buttonStyles) {
    return (
      <TouchableOpacity
        style={{ ...styles.actionButton, ...buttonStyles }}
        onPress={onPress}
      >
        <Image style={styles.actionButtonIcon} source={src} />
      </TouchableOpacity>
    );
  }

  toggleFollowModal = (x = !this.state.followModal) => {
    this.setState({
      followModal: x,
      accepted: false,
    });
  };

  getProfilePicture() {
    if (this.props.notification_type == "status") {
      let profilePicture = this.props.data.user.profile_picture;
      return profilePicture
        ? { url: profilePicture }
        : require("../assets/empty-profile-picture.png");
    } else if (this.props.notification_type == "relationship") {
      let profilePicture = this.props.data.follower.profile_picture;
      return profilePicture
        ? { url: profilePicture }
        : require("../assets/empty-profile-picture.png");
    } else {
      return require("../assets/empty-profile-picture.png");
    }
  }

  getFollowModal = () => {
    const connection = {
      startTime: this.props.data.startTime,
      intent: this.props.data.intent || "",
      message: this.props.data.message || "",
    };

    return (
      <FollowRequestModal
        visible={this.state.followModal}
        toggle={this.toggleFollowModal}
        accepted={this.state.accepted}
        data={this.props.data}
        connection={connection}
        respond={this.respond}
        profilePicture={this.getProfilePicture()}
        onModalHide={() => {}}
        connectModal={
          <ConnectModal
            visible={this.state.connectModal && this.state.accept}
            accepted={this.state.accepted}
            toggleModal={this.toggleFollowModal}
            respond={this.addToCircle}
          />
        }
      />
    );
  };

  getContent = () => {
    switch (this.props.notification_type) {
      case "relationship": {

        return (
          <TouchableOpacity
            style={styles.textContainer}
            onPress={() =>
              this.props.data.status == "PENDING"
                ? this.toggleFollowModal()
                : null
            }
            onLongPress={() => this.props.deleteNotification(this.props.id)}
          >
            {this.getFollowModal()}
            <View style={styles.outline}>
              <Image style={styles.image} source={this.getProfilePicture()} />
              <View style={styles.text}>
                <View style={{ flexDirection: "row" }}>
                  <Text numberOfLines={1} style={styles.notificationText}>
                    {" "}
                    {this.props.text}{" "}
                  </Text>
                </View>
                <Text style={styles.subtitle}>
                  {timeAgo.format(parseInt(this.props.timestamp))}
                </Text>
              </View>
              {this.getRequestButtons()}
            </View>
          </TouchableOpacity>
        );
      }
      case "status": {
      }
      default: {
        return (
          <TouchableOpacity
            style={styles.textContainer}
            onLongPress={() => this.props.deleteNotification(this.props.id)}
          >
            <View style={styles.outline}>
              <Image style={styles.image} source={this.getProfilePicture()} />
              <View style={styles.text}>
                <View style={{ flexDirection: "row" }}>
                  <Text numberOfLines={1} style={styles.notificationText}>
                    {" "}
                    {this.props.text}{" "}
                  </Text>
                </View>
                <Text style={styles.subtitle}>
                  {timeAgo.format(parseInt(this.props.timestamp))}
                </Text>
              </View>
              {this.getRequestButtons()}
            </View>
          </TouchableOpacity>
        );
      }
    }
  };

  render() {
    return (
      <View style={{ flexDirection: "row", ...styles.container }}>
        {this.getContent()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    borderColor: "grey",
    borderBottomWidth: 1,
    height: 70,
    backgroundColor: "white",
  },
  textContainer: {
    flexDirection: "row",
    flex: 1,
  },
  view: {
    alignItems: "center",
    flexDirection: "row",
    backgroundColor: "white",
    width: "100%",
    height: "10%",
    padding: 18,
    marginBottom: 1,
  },
  requestButtons: {
    position: "absolute",
    marginRight: 10,
    right: 0,
  },
  text: {
    flexDirection: "column",
    left: "auto",
    marginLeft: 5,
  },
  title: {
    fontFamily: "poppins-semi-bold",
    fontSize: 10,
    marginTop: 3,
  },
  notificationText: {
    fontWeight: "normal",
    fontFamily: "poppins-semi-bold",
    fontSize: 9,
    marginTop: 4,
  },
  subtitle: {
    fontWeight: "normal",
    fontSize: 10,
    color: "grey",
    fontFamily: "poppins-light",
  },
  outline: {
    alignItems: "center",
    justifyContent: "flex-start",
    flexDirection: "row",
    flex: 1,
    paddingLeft: 10,
  },
  image: {
    borderRadius: 150 / 2,
    backgroundColor: "white",
    width: 40,
    height: 40,
    overflow: "hidden",
  },
  actionButtonIcon: {
    width: 18,
    height: 18,
    overflow: "hidden",
    marginLeft: 7,
    marginTop: 7,
  },
  actionButton: {
    borderRadius: 150 / 2,
    width: 33,
    height: 33,
    overflow: "hidden",
  },
});

export default NotificationItemComponent;
