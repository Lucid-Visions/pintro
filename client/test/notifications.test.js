import React from "react";
import { create, act } from "react-test-renderer";
import NotificationItem from "../src/components/NotificationItem";

// mock fetch functions
global.fetch = jest.fn()
global.env = jest.fn()
// set test `Headers`
global.Headers = () => ({
  append: jest.fn(),
});

/**
 * Tests for the Notification Item of type relationship
 */
describe("<NotificationItem type={`relationship`}/>", () => {
  let notificationDeleted = false;
  let deletedNotificationId;
  let feedItems = 2;

  let props;
  const createProps = () => ({
    key: 1,
    text: "Kim wants to connect with you",
    timestamp: 15864314430000,
    notification_type: "relationship",
    data: {
      status: "PENDING",
      follower: {
        _id: "431",
        name: "Kim",
        profile_picture: null,
        bio: "Kim is a very good app tester",
      },
    },
    id: "12e1ce12e3",
    deleteNotification: jest.fn((id) => {
      deletedNotificationId = id;
      notificationDeleted = true;
      feedItems -= 1;
    }),
    refresh: jest.fn(),
  });

  beforeEach(() => {
    props = createProps();
  });

  test("snapshot of the notification item component", async () => {
    const feed = create(<NotificationItem {...props} />).toJSON();
    expect(feed).toMatchSnapshot();
  });

  test("test if onLongPress will delete the notification", async () => {
    const component = create(<NotificationItem {...props} />);
    const instance = component.root;
    const notification = instance.findByType("View");
    const notificationTouchableOpacity = notification.props.children.props;

    expect(notificationDeleted).toBeFalsy();
    expect(feedItems).toBe(2)
    notificationTouchableOpacity.onLongPress();
    expect(notificationDeleted).toBe(true);
    expect(deletedNotificationId).toBe(props.id);
    expect(feedItems).toBe(1)
  });

  test("test if FollowRequestModal will open onPress the notification item", () => {
    const component = create(<NotificationItem {...props} />);
    const instance = component.root;
    const notification = instance.findByType("View");
    const notificationTouchableOpacity = notification.props.children.props;

    expect(instance.instance.state.followModal).toBeFalsy();
    notificationTouchableOpacity.onPress()
    expect(instance.instance.state.followModal).toBe(true);
  })

  test("test decline request button", () => {
    const component = create(<NotificationItem {...props} />);
    const instance = component.root;
    const notification = instance.findByType("View");
    const declineButton = notification.props.children.props.children[1].props.children[2].props.children[0].props;

    declineButton.onPress()
    expect(instance.instance.state.accepted).toBeFalsy();
  })

  test("test accept request button", () => {
    const component = create(<NotificationItem {...props} />);
    const instance = component.root;
    const notification = instance.findByType("View");
    const acceptButton = notification.props.children.props.children[1].props.children[2].props.children[1].props;

    expect(instance.instance.state.accepted).toBeFalsy();
    acceptButton.onPress()
    expect(instance.instance.state.accepted).toBe(true);
  })
});

/**
 * Tests for the Notification Item of type status
 */
describe("<NotificationItem type={`status`}/>", () => {
  let notificationDeleted = false;
  let deletedNotificationId;
  let feedItems = 2;

  let props;
  const createProps = () => ({
    key: 1,
    text: "Kim liked your status",
    timestamp: 15864314430000,
    notification_type: "status",
    id: "12e1ce12e3",
    data: {
      user: {
        _id: "431",
        name: "Kim",
        profile_picture: null,
        bio: "Kim is a very good app tester",
      },
    },
    deleteNotification: jest.fn((id) => {
      deletedNotificationId = id;
      notificationDeleted = true;
      feedItems -= 1;
    }),
    refresh: jest.fn(),
  });

  beforeEach(() => {
    props = createProps();
  });

  test("snapshot of the notification item component", async () => {
    const feed = create(<NotificationItem {...props} />).toJSON();
    expect(feed).toMatchSnapshot();
  });

  test("test if onLongPress will delete the notification", async () => {
    const component = create(<NotificationItem {...props} />);
    const instance = component.root;
    const notification = instance.findByType("View");
    const notificationTouchableOpacity = notification.props.children.props;

    expect(notificationDeleted).toBeFalsy();
    expect(feedItems).toBe(2)
    notificationTouchableOpacity.onLongPress();
    expect(notificationDeleted).toBe(true);
    expect(deletedNotificationId).toBe(props.id);
    expect(feedItems).toBe(1)
  });
})
