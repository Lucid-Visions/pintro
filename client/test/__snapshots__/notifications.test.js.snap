// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationItem type={\`relationship\`}/> snapshot of the notification item component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 1,
      "borderColor": "grey",
      "flex": 1,
      "flexDirection": "row",
      "height": 70,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onRequestClose={[Function]}
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeThreshold={100}
      transparent={true}
      visible={false}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "black",
            "bottom": 0,
            "height": 1334,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 750,
          }
        }
      />
      <View
        hideModalContentWhileAnimating={false}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        pointerEvents="box-none"
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#FCFCFC",
            "borderRadius": 20,
            "flex": 1,
            "height": 333.5,
            "justifyContent": "space-evenly",
            "margin": 37.5,
            "marginTop": 296.44444444444446,
            "maxHeight": 350,
            "transform": Array [
              Object {
                "translateY": 0,
              },
            ],
            "width": 681.8181818181818,
          }
        }
        supportedOrientations={
          Array [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
      >
        <View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "width": 625,
              }
            }
          >
            <Image
              source={1}
              style={
                Object {
                  "alignSelf": "center",
                  "borderRadius": 30,
                  "height": 60,
                  "overflow": "hidden",
                  "width": 60,
                }
              }
            />
            <View
              style={
                Object {
                  "marginLeft": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "black",
                    "fontFamily": "poppins-bold",
                    "fontSize": 13,
                    "margin": "auto",
                    "textAlign": "left",
                  }
                }
              >
                Kim
              </Text>
              <Text
                style={
                  Object {
                    "color": "black",
                    "fontFamily": "poppins-medium",
                    "fontSize": 13,
                    "textAlign": "left",
                    "width": 200,
                  }
                }
              >
                Kim is a very good app tester...
              </Text>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "marginBottom": 5,
                    "marginTop": 5,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 13,
                      "marginRight": 5,
                      "resizeMode": "contain",
                      "width": 10,
                    }
                  }
                />
                <Text
                  style={
                    Object {
                      "color": "black",
                      "fontFamily": "poppins-light",
                      "fontSize": 11,
                      "margin": "auto",
                      "textAlign": "left",
                    }
                  }
                >
                   
                  in NaN years
                   
                </Text>
              </View>
            </View>
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            >
              <Image
                source={1}
                style={
                  Object {
                    "height": 13,
                    "marginTop": 4,
                    "width": 13,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "alignSelf": "flex-start",
                    "color": "#F8AE39",
                  }
                }
              >
                 
                 
                km
                 
              </Text>
            </View>
          </View>
          <View
            flexDirection="row"
            style={
              Object {
                "alignItems": "space-between",
                "borderColor": "orange",
                "borderWidth": 0,
                "justifyContent": "space-between",
                "marginTop": 20,
                "width": 625,
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "black",
                  "borderColor": "black",
                  "borderRadius": 7.5,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 35,
                  "justifyContent": "center",
                  "opacity": 1,
                  "width": "47%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontFamily": "poppins-medium",
                    "fontSize": 10,
                  }
                }
              >
                ACCEPT
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderColor": "black",
                  "borderRadius": 7.5,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 35,
                  "justifyContent": "center",
                  "opacity": 1,
                  "width": "47%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "black",
                    "fontFamily": "poppins-medium",
                    "fontSize": 10,
                  }
                }
              >
                DECLINE
              </Text>
            </View>
          </View>
        </View>
      </View>
    </Modal>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "paddingLeft": 10,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 75,
            "height": 40,
            "overflow": "hidden",
            "width": 40,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "column",
            "left": "auto",
            "marginLeft": 5,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "fontFamily": "poppins-semi-bold",
                "fontSize": 9,
                "fontWeight": "normal",
                "marginTop": 4,
              }
            }
          >
             
            Kim wants to connect with you
             
          </Text>
        </View>
        <Text
          style={
            Object {
              "color": "grey",
              "fontFamily": "poppins-light",
              "fontSize": 10,
              "fontWeight": "normal",
            }
          }
        >
          in 452 years
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginRight": 10,
            "position": "absolute",
            "right": 0,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "borderRadius": 75,
              "borderWidth": 1,
              "height": 33,
              "opacity": 1,
              "overflow": "hidden",
              "width": 33,
            }
          }
        >
          <Image
            source={1}
            style={
              Object {
                "height": 18,
                "marginLeft": 7,
                "marginTop": 7,
                "overflow": "hidden",
                "width": 18,
              }
            }
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": 75,
              "height": 33,
              "marginLeft": 12,
              "opacity": 1,
              "overflow": "hidden",
              "width": 33,
            }
          }
        >
          <Image
            source={1}
            style={
              Object {
                "height": 18,
                "marginLeft": 7,
                "marginTop": 7,
                "overflow": "hidden",
                "width": 18,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`<NotificationItem type={\`status\`}/> snapshot of the notification item component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 1,
      "borderColor": "grey",
      "flex": 1,
      "flexDirection": "row",
      "height": 70,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "paddingLeft": 10,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 75,
            "height": 40,
            "overflow": "hidden",
            "width": 40,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "column",
            "left": "auto",
            "marginLeft": 5,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "fontFamily": "poppins-semi-bold",
                "fontSize": 9,
                "fontWeight": "normal",
                "marginTop": 4,
              }
            }
          >
             
            Kim liked your status
             
          </Text>
        </View>
        <Text
          style={
            Object {
              "color": "grey",
              "fontFamily": "poppins-light",
              "fontSize": 10,
              "fontWeight": "normal",
            }
          }
        >
          in 452 years
        </Text>
      </View>
    </View>
  </View>
</View>
`;
